### ISSUE TITLE
Proposal: fix ImageDelete to only prevent untag of references that are used

### ISSUE THREAD

**Issue by dnephin** [moby/moby#36435]
Original discussion in #36295

Currently `ImageDelete` will untag references as long as there remains at least 1 reference to the image id, even if that reference is used by a container or service. It refuses to untag a reference if there is at least one container using the image id referenced by the tag, even if that tag is not used by any container or service.

I propose the behaviour should be changed so that `ImageDelete` will untag of any references that are not being used. Image tags that are referenced can still be removed with `force`. This behaviour is easier to explain and is more intuitive.

An example of the unexpected behaviour:

```
$ docker pull redis:alpine
...
$ docker tag redis:alpine myimage:1
$ docker run -d myimage:1
552ae8f1134235f9c7d41ec220c9978cb51fd387d4a2d2cd4215b108950ee7c1
$ docker container ls --format '{{.Image}}'
myimage:1
$ docker image rm myimage:1
Untagged: myimage:1
$ docker image rm redis:alpine
Error response from daemon: conflict: unable to remove repository reference "redis:alpine" (must force) - container 20c7d341f7f5 is using its referenced image d3117424aaee
$ docker container ls --format '{{.Image}}'
d3117424aaee
```

`ImageDelete` allows `myimage:1` to be removed (even through it is referenced), but doesn't allow me to remove (unless forced) `redis:alpine` which is not referenced by any containers.

**Comment by thaJeztah** [comment]
:+1: current behaviour is confusing; most people will see the "tag" as the actual "image", so refusing to remove a tag if it's in use will better match that expectation.

/cc @tianon @tonistiigi

**Comment by thaJeztah** [comment]
If the reason is to prevent having no references to the image, it's also inconsistent:

Pull an old version of an image and tag it "latest"

```bash
docker image pull busybox:1.27
docker image tag busybox:1.27 busybox:latest

docker image ls

REPOSITORY                TAG                 IMAGE ID            CREATED             SIZE
busybox                   1.27                6ad733544a63        3 months ago        1.13MB
busybox                   latest              6ad733544a63        3 months ago        1.13MB
```

Start a container using the old image:

```bash
docker container run -dit busybox:1.27
```

Remove the image (allowed):

```bash
docker image rm busybox:1.27
Untagged: busybox:1.27
```

Remove the _unused_ `:latest` tag (not allowed):

```bash
docker image rm busybox:latest
Error response from daemon: conflict: unable to remove repository reference "busybox:latest" (must force) - container 66082b0c7fd0 is using its referenced image 6ad733544a63
```

Pull an updated version of the `:latest` image:

```
docker image pull busybox:latest

busybox                   latest              5b0d59026729        5 weeks ago         1.15MB
busybox                   <none>              6ad733544a63        3 months ago        1.13MB
```

Now remove the image:

```bash
docker image rm busybox:latest

Untagged: busybox:latest
Untagged: busybox@sha256:1669a6aa7350e1cdd28f972ddad5aceba2912f589f19a090ac75b7083da748db
Untagged: busybox@sha256:187c294487772d71eb5e6675f31e8863fcd655e44f634ffbc92ed30f407a1b9d
```

**Comment by tonistiigi** [comment]
This is a breaking change, most notably breaking all scripts that currently do the container upgrade to a new image with renaming the old image for rollback compatibility. Using force with a new meaning to untag a referenced image is also a breaking change(or if you change it to remove the image there then you break all people using this method for untagging). The proposal does not define what happens when last "unreferenced" tag is removed, assuming it only removes the tag it is a third breaking change where currently force removing an image with a single tag is guaranteed to remove the image if it returns successfully. Not to mention, this completely changes the behavior of all commands for people who start the images by the image id/prefix, digest etc., currently the method used to refer to an image in container startup has no meaning for any of the next commands.

It does not fix the inconsistencies for the services as the reference used to start a service is different from the reference starting a container(one is expanded).

@thaJeztah Source of that confusion is more in the view that automatically tries to add digest references on pull, and then not even show them on `image ls` by default. Last part is consistent as far as `pull/rm` go. If a pull was successful and actually pulled a new image then deleting it in subsequent command should always succeed as well.

**Comment by dnephin** [comment]
> most notably breaking all scripts that currently do the container upgrade to a new image with renaming the old image for rollback compatibility

This is not broken:

```
docker tag image image-backup
docker pull image
docker stop container
docker run container
```

Could you elaborate on why `image rm` is needed for this?

> Using force with a new meaning to untag a referenced image is also a breaking change

```
$ docker image rm alpine:3.5
Error response from daemon: conflict: unable to remove repository reference "alpine:3.5" (must force) - container 3c559c7b5871 is using its referenced image 6c6084ed97e5
$ docker image rm -f alpine:3.5
Untagged: alpine:3.5
Untagged: alpine@sha256:9148d069e50eee519ec45e5683e56a1c217b61a52ed90eb77bdce674cc212f1e
```

That is the current behaviour today, so how is it a breaking change?

> The proposal does not define what happens when last "unreferenced" tag is removed, assuming it only removes the tag 

This proposal doesn't define that behaviour because it will not change. The behaviour will continue to be untag and remove.

> this completely changes the behavior of all commands for people who start the images by the image id/prefix,... currently the method used to refer to an image in container startup has no meaning for any of the next commands.

I'm not sure what you mean by this. Could you elaborate? The behaviour of other commands should not change.

> It does not fix the inconsistencies for the services

Not related to this proposal. This proposal is only meant to fix the problem with `image remove`.

**Comment by tonistiigi** [comment]
> This is not broken:

I didn't mean that there aren't workarounds to fix these scripts, just that everything that currently does an action of "renaming a tag" will start to give errors. I agree that in most cases they can be modified to use some other command and have the same behavior, or like in your example skip rm if you know the next step is to pull it. To do a change like that we would need a long deprecation period with warnings on affected cases and even then we will still break automation. So to justify a big change in documented behavior that has existed for ~4 years we would need a big outcry from the community. The issue that started this discussion wasn't even vaguely related to any of these commands.

> That is the current behavior today, so how is it a breaking change?

This is not the behavior on stopped containers that all the other examples were about.

> This proposal doesn't define that behaviour because it will not change. The behaviour will continue to be untag and remove.

If the proposed behavior really is to "untag and remove" it means that you can have containers that do not point to any image, without ever using the force flag, what was exactly what the `force` flag was trying to avoid.

> I'm not sure what you mean by this. Could you elaborate? The behaviour of other commands should not change.

With your proposal depending on if the user runs `docker run busybox ls` or `docker run $(docker inspect -f '{{.ID}}' busybox) ls` it changes the outcome of `docker rmi busybox`. A lot of tooling (and copy-paste cases) definitely use the latter because these are the values other API commands return.

**Comment by dnephin** [comment]
> everything that currently does an action of "renaming a tag" will start to give errors

I'm still trying to understand this argument. To me "renaming a tag" has *nothing* to do with `image rm`. Why would you ever need to rm something before replacing it, instead of just replacing it? 

> With your proposal depending on if the user runs `docker run busybox ls` or `docker run $(docker inspect -f '{{.ID}}' busybox) ls` it changes the outcome of `docker rmi busybox`

As it should. Those are different commands. One expresses intent to run busybox. The other expresses into to run a specific image id.

**Comment by tonistiigi** [comment]
Yes, in an ideal world renaming a tag shouldn't have anything to with `image rm` as we would have a `untag` command to handle that. But we don't, and at least I don't know any other way to achieve that (without creating some dummy images) without `rmi`. Saying that all people should always just replace images with something else is as good as saying that we don't need a way to remove tags at all.

> One expresses intent to run busybox. The other expresses into to run a specific image id.

But busybox is the image, there is only one object in the API. You specify an image to start a container, "busybox" is one of the strings in the "repotags" field of that image. Again, I'm just pointing out all the cases this would break, not arguing what would be a better compromise if we were to design it from scratch.

**Comment by thaJeztah** [comment]
> most notably breaking all scripts that currently do the container upgrade to a new image with renaming the old image for rollback compatibility

I'm also having trouble grasping this scenario; with "renaming the old image" do you mean something like:

```
docker run -dit busybox:latest
7ab7336b325c35cc9502ab0a02462b065d731dd01a6509311c4b0a5530eabd84

docker tag busybox:latest busybox:previous
docker rmi busybox:latest
Untagged: busybox:latest

docker pull busybox:latest
latest: Pulling from library/busybox
Digest: sha256:2107a35b58593c58ec5f4e8f2c4a70d195321078aebfadfbfb223a2ff4a4ed21
Status: Downloaded newer image for busybox:latest
```

Also not sure I understand the container "upgrade" (as the container wouldn't change by those steps?)

**Comment by tonistiigi** [comment]
@thaJeztah Only the second step in your example is about renaming a tag on an image. A pull doesn't have anything to do with it. You can rename just because you found a typo, or for whatever reason.

> as the container wouldn't change by those steps

Not discussing any actions on new containers. The case is for stopping a container (maybe renaming a container itself), not deleting it, renaming an image that container used to `myimage.backup.yymmdd` in case there is ever a need to roll back to that container (and maybe restore the backup to its original name).

**Comment by dnephin** [comment]
I think the important thing is that the rename (`docker tag`) still works with this proposal.

What isn't clear is why you need to `image rm` in that case. That is an unnecessary action with no benefit (that I can see). What is the value in removing a tag that is still being used?

The scenario you describe doesn't require the tag to be removed. It only requires that a new tag is copied, and that some build or pull operation can override the old tag, both of which will still be supported.
